### PR TITLE
ARROW-12134: [C++] Add match_substring_regex kernel

### PR DESCRIPTION
**Pull request by lidavidm**
For consistency with match_substring, this is the equivalent of Python's re.search(), not re.match().

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 1]
```diff
@@ -411,40 +411,104 @@ void TransformMatchSubstring(const uint8_t* pattern, int64_t pattern_length,
...
 
-template <typename Type>
+template <typename Type, template <typename offset_type> class Matcher>
 struct MatchSubstring {
   using offset_type = typename Type::offset_type;
   static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
-    MatchSubstringOptions arg = MatchSubstringState::Get(ctx);
-    const uint8_t* pat = reinterpret_cast<const uint8_t*>(arg.pattern.c_str());
-    const int64_t pat_size = arg.pattern.length();
+    // TODO Cache matcher accross invocations (for regex compilation)
```
Comment (cyb70289) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:418]: `accross` -> `across`?

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 2]
```diff
@@ -411,40 +411,104 @@ void TransformMatchSubstring(const uint8_t* pattern, int64_t pattern_length,
...
+    auto exec_32 = MatchSubstring<StringType, PlainSubstringMatcher>::Exec;
+    auto exec_64 = MatchSubstring<LargeStringType, PlainSubstringMatcher>::Exec;
+    DCHECK_OK(func->AddKernel({utf8()}, boolean(), exec_32, MatchSubstringState::Init));
+    DCHECK_OK(
+        func->AddKernel({large_utf8()}, boolean(), exec_64, MatchSubstringState::Init));
+    DCHECK_OK(registry->AddFunction(std::move(func)));
+  }
+#ifdef ARROW_WITH_RE2
+  {
+    auto func = std::make_shared<ScalarFunction>("match_substring_regex", Arity::Unary(),
```
Comment (cyb70289) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:502]: New functions need be documented in https://github.com/apache/arrow/blob/master/docs/source/cpp/compute.rst

@pitrou , should we also document new functions in python api document?
https://github.com/apache/arrow/blob/master/docs/source/python/api/compute.rst

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:502]: Ah, good point! We should indeed add the names there to auto-generate the API docs. I keep forgetting about it, so there's probably more to add.

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 3]
```diff
@@ -411,40 +411,104 @@ void TransformMatchSubstring(const uint8_t* pattern, int64_t pattern_length,
...
+             int64_t output_offset, uint8_t* output) {
+    const uint8_t* pat = reinterpret_cast<const uint8_t*>(options_.pattern.c_str());
+    const int64_t pat_size = options_.pattern.length();
+    TransformMatchSubstring<offset_type>(pat, pat_size, offsets, data, length,
+                                         output_offset, output);
+  }
+};
+
+template <typename offset_type>
+struct RegexSubstringMatcher {
```
Comment (cyb70289) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:449]: Should this struct wrapped by `#ifdef ARROW_WITH_RE2`?

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:456]: You are right, good catch.

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 4]
```diff
@@ -411,40 +411,106 @@ void TransformMatchSubstring(const uint8_t* pattern, int64_t pattern_length,
...
+};
+
 const FunctionDoc match_substring_doc(
     "Match strings against literal pattern",
     ("For each string in `strings`, emit true iff it contains a given pattern.\n"
      "Null inputs emit null.  The pattern must be given in MatchSubstringOptions."),
     {"strings"}, "MatchSubstringOptions");
 
+#ifdef ARROW_WITH_RE2
+template <typename offset_type>
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:455]: There's no conceptual reason for matchers to be templated on offset type. The iteration loop on string values could be moved out of the matcher and into the calling `MatchSubstring` class. Then all a matcher does is take a `string_view` or similar and return the index where the pattern was found (or presumably -1 if not found).

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/scalar_string.cc (hunk 4 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_string.cc]: Got it, I refactored that part. The Knuth-Morris-Pratt implementation also had to be refactored as it was templated on offset_type, though it seems that was unnecessary.

File: docs/source/cpp/compute.rst [hunk 5]
```diff
@@ -522,26 +522,31 @@ These functions trim off characters on both sides (trim), or the left (ltrim) or
...
-|                    |            | Binary- and String-like            |               |                                        |
-+--------------------+------------+------------------------------------+---------------+----------------------------------------+
++---------------------------+------------+------------------------------------+---------------+----------------------------------------+
+| Function name             | Arity      | Input types                        | Output type   | Options class                          |
++===========================+============+====================================+===============+========================================+
+| match_substring           | Unary      | String-like                        | Boolean (1)   | :struct:`MatchSubstringOptions`        |
++---------------------------+------------+------------------------------------+---------------+----------------------------------------+
+| match_substring_regex     | Unary      | String-like                        | Boolean (1)   | :struct:`MatchSubstringOptions`        |
```
Comment (pitrou) [inline docs/source/cpp/compute.rst:530]: You missed the renumbering of notes here (and below).

**Review by pitrou — CHANGES_REQUESTED** [review]
LGTM, one last nit

**Review by cyb70289 — APPROVED** [review]
~~LGTM, +1~~

Oops, forget to check the CI.

**Comment by pitrou** [comment]
It looks like there's a failure on Windows in debug mode (see CI):
```
[ RUN      ] TestStringKernels/0.MatchSubstring
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29910\include\vector(1508) : Assertion failed: vector subscript out of range
```

**Comment by lidavidm** [comment]
It's reproducible on Linux by temporarily using .at() for array access. I incorrectly turned `vector<T> foo(N)` into `reserve(N)` when it should have been `resize(N)`.

**Comment by lidavidm** [comment]
This should be ready again (minus RTools 3.5/JNI which are known to be failing/flaky).

**Review by pitrou — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**